dohproxy/server_protocol.py

```diff
@@ -42,10 +43,34 @@ def __init__(self, upstream_resolver, upstream_port, logger=None):
         self.logger = logger
         self.transport = None
 
-    async def query(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
-        dnsr = await self.query_udp(dnsq, clientip, timeout=timeout)
+    async def query(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT,
+                    ecs=False):
+        original_ecs_option = None
+        for option in dnsq.options:
+            if isinstance(option, dns.edns.ECSOption):
+                original_ecs_option = option
+                break
+
+        # (Potentially) modified copy of dnsq
+        dnsq_mod = dns.message.from_wire(dnsq.to_wire())
+        if ecs:
+            utils.set_dns_ecs(dnsq_mod, clientip)
+
+        dnsr = await self.query_udp(dnsq_mod, clientip, timeout=timeout)
         if dnsr is None or (dnsr.flags & dns.flags.TC):
-            dnsr = await self.query_tcp(dnsq, clientip, timeout=timeout)
+            dnsr = await self.query_tcp(dnsq_mod, clientip, timeout=timeout)
+
+        if dnsr is not None and ecs:
+            options = []
+            for option in dnsr.options:
+                if isinstance(option, dns.edns.ECSOption):
+                    if original_ecs_option is not None:
+                        options.append(original_ecs_option)
```

Comment:
this would be invalid as it would put an ECS query option in the response.
I think the logic would be something along:
```
if original_ecs_option is None:
  pop_ecs_from_response()
```

e.g, if we were not provided an ECS optiopn in the query, we would take it out of the response, otherwise we leave it untouched.

Reply:
Thanks.  That logic was from earlier on when trying to deal with a scenario where the client had ECS set but we were modifying ECS anyway, which of course we don't actually want to do.  I was able to simplify the logic now and not need to track the original ECS option, just by passing from set_dns_ecs() whether our ECS was actually set or not.  The logic should be sound now.